packages/cloudflare-access/functions/_middleware.ts

```diff
@@ -8,38 +7,102 @@ type CloudflareAccessPagesPluginFunction<
   Data extends Record<string, unknown> = Record<string, unknown>
 > = PagesPluginFunction<Env, Params, Data, PluginArgs>;
 
-const extractJWTFromRequest = (request: Request) => {
-  return request.headers.get("Cf-Access-Jwt-Assertion");
+const extractJWTFromRequest = (request: Request) =>
+  request.headers.get("Cf-Access-Jwt-Assertion");
+
+// Adapted slightly from https://github.com/cloudflare/workers-access-external-auth-example
+const base64URLDecode = (s: string) => {
+  s = s.replace(/-/g, "+").replace(/_/g, "/").replace(/\s/g, "");
+  return new Uint8Array(
+    Array.prototype.map.call(atob(s), (c: string) => c.charCodeAt(0))
+  );
+};
+
+const asciiToUint8Array = (s: string) => {
+  let chars = [];
+  for (let i = 0; i < s.length; ++i) {
+    chars.push(s.charCodeAt(i));
+  }
+  return new Uint8Array(chars);
 };
 
 const generateValidator =
   ({ domain, aud }: { domain: string; aud: string }) =>
   async (
     request: Request
-  ): Promise<false | { jwt: string; payload: object }> => {
+  ): Promise<{
+    jwt: string;
+    payload: object;
+  }> => {
     const jwt = extractJWTFromRequest(request);
 
-    const { payloadObj, headerObj } = KJUR.jws.JWS.parse(jwt);
-    const { kid } = headerObj as KJUR.jws.JWS.JWSResult["headerObj"] & {
-      kid: string;
-    };
+    const parts = jwt.split(".");
+    if (parts.length !== 3) {
+      throw new Error("JWT does not have three parts.");
+    }
+    const [header, payload, signature] = parts;
+
+    const textDecoder = new TextDecoder("utf-8");
+    const { kid, alg, typ } = JSON.parse(
+      textDecoder.decode(base64URLDecode(header))
+    );
+    if (typ !== "JWT" || alg !== "RS256") {
+      throw new Error("Unknown JWT type or algorithm.");
+    }
 
     const certsURL = new URL("/cdn-cgi/access/certs", domain);
     const certsResponse = await fetch(certsURL.toString());
-    const { public_certs } = (await certsResponse.json()) as {
-      keys: Record<string, string>[];
+    const { keys } = (await certsResponse.json()) as {
+      keys: ({
+        kid: string;
+      } & JsonWebKey)[];
       public_cert: { kid: string; cert: string };
       public_certs: { kid: string; cert: string }[];
     };
-    const cert = public_certs.find(({ kid: id }) => id === kid).cert;
+    if (!keys) {
+      throw new Error("Could not fetch signing keys.");
+    }
+    const jwk = keys.find((key) => key.kid === kid);
+    if (!jwk) {
+      throw new Error("Could not find matching signing key.");
+    }
+    if (jwk.kty !== "RSA" || jwk.alg !== "RS256") {
+      throw new Error("Unknown key type of algorithm.");
+    }
 
-    const isValid = KJUR.jws.JWS.verifyJWT(jwt, cert, {
-      alg: ["RS256"],
-      aud: [aud],
-    });
+    const key = await crypto.subtle.importKey(
+      "jwk",
+      jwk,
+      { name: "RSASSA-PKCS1-v1_5", hash: "SHA-256" },
+      false,
+      ["verify"]
+    );
 
-    if (!isValid) {
-      return false;
+    const verified = await crypto.subtle.verify(
```

Comment:
Nitpick: First do the inexpensive checks before doing the expensive (crypto) checks